spectre-and-meltdown-on-apu2.md: initial version of the post (#43)
3mdeb/news-and-ideas: blog/content/post/2019-05-29-spectre-and-meltdown-on-apu2.md

---
title: Meltdown and Spectre on PC Engines apu2
abstract: As a continuation the Meltdown and Spectre blog post, this post
          present the vulnerability status and mitigation with microcode
          update on PC Engines apu2. Read the post and get to know the open
          source tools for vulnerability and mitigation checks, as well as
          exploiting proof of concepts.
cover: /covers/spectre_apu2.png
author: Michał.Żygowski
layout: post
published: true
date: 2019-05-29

tags:
    - spectre
    - meltdown
    - apu2
    - PC Engines
    - coreboot
categories:
    - Firmware
    - Security

---

# Meltdown and Spectre

In the [previous post](https://blog.3mdeb.com/2019/2019-03-14-meltdown-and-spectre-vulnerabilities/)
I have introduced the Meltdown and Spectre vulnerabilities of modern x86
processors and what threat do they pose to security and safety of the data. As
a continuation of the last post, I will demonstrate the state of Meltdown and

> Spectre vulnerabilities on PC Engines apu2 platforms. We have recently added a

Reviewer: @miczyg1 recently?
```suggestion
Spectre vulnerabilities on PC Engines apu2 platforms. Some time ago we have added a
```

Author: Done

microcode update feature to PC Engines firmware so I will show how microcode

> update improves mitigation and present results using public tools/Proof of

Reviewer: ```suggestion
update improves mitigation and present results using public tools/proof of
```

Author: Done

Reading at malicious_x = 0xffffffffffdfeede... Success: 0xFF=’?’ score=0 
Reading at malicious_x = 0xffffffffffdfeedf... Success: 0xFF=’?’ score=0 
```

One can see that positive results are obtained with a larger value of cache hit
threshold (with 70 some characters have not been obtained - score 0, 100 is sufficient
to get whole string - score 2 everywhere). It is related to CPU performance.

> For those more interested in details of the attack, please refer to
> [Spectre Attacks](https://spectreattack.com/spectre.pdf) or to my
> [previous post](https://blog.3mdeb.com/2019/2019-03-14-meltdown-and-spectre-vulnerabilities/)
> introducing the Meltdown and Spectre.

Let's also check out the kernel mitigation. Rebuild the PoC with:

```shell
make clean
CFLAGS=-DLINUX_KERNEL_MITIGATION make
```

After trying to run the PoC with 20/70/100 the string is not disclosed. But
what if the kernel we use does not have the mitigation? The answer is firmware
and microcode for full hardware mitigation.

# Microcode update and spectre

To present results of Spectre vulnerability without kernel mitigation, we will
need the microcode update. For details how to build PC Engines apu2 coreboot

> firmware with microcode update, please refer to this [guide](https://github.com/pcengines/apu2-documentation/blob/master/docs/microcode_patching.md)

Reviewer: @miczyg1 does this procedure still work fine? There are some bug reports related: https://github.com/pcengines/apu2-documentation/issues/145

Author: @pietrushnic this bug is related to docker itself not to build process. I have an asciinema in the pst showing how to build the image. Worked perfectly fine for the version mentioned in blogpost.

Author: Also checked with new v4.9.0.5. No problems with building and flashing. For mainline releases we use unmodified coreboot-sdk container, so it is the Linxu configuration problem with permissions. It is highly dependable on groups, directory configuration, permissions, structure. May even be a file system (like if someone would like to compile coreboot on NTFS for some reason). Anyway, I have built is successfully without problems, waiting for the response to my comment in linked issue.

Reviewer: Ok

Used scripts and tools give only overall insight into the status of the
Meltdown and Spectre. They are not official tools proving system vulnerability,
they were designed to give information on present mitigation (spectre meltdown
checker) or to try exploiting the vulnerabilities based on an [article](https://spectreattack.com/spectre.pdf).

It has to be noted that AMD and Intel processors are affected differently. AMD
processors are "marked" to not be vulnerable for some variants. AMD has even
released a [statement regarding the speculative execution.](https://developer.amd.com/wp-content/resources/Managing-Speculation-on-AMD-Processors.pdf)

During experiments, different results have been obtained. It proves that only
combined mitigation in firmware/microcode and kernel give the best protection.
Although microcode update seems to not improve the situation in this particular
case.

I hope this post was useful for you. Please feel free to share your opinion and
if you think there is value, then share with friends.

If you think we can help in improving the security of your firmware or you
looking for someone who can boot your product by leveraging advanced features

> of used hardware platform, feel free to [boot a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)

Reviewer: ```suggestion
of used hardware platform, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
```

Author: Done
